src/nlsolve/functional.jl

```diff
@@ -1,154 +1,291 @@
-      z₊ = dt .* f(u, p, tstep) .- mz .+ z
+      dz = dt .* f(u, p, tstep) .- mz
+      z₊ = z .+ dz
     end
 
-    # check early stopping criterion
-    ndzprev = ndz
-    ndz = integrator.opts.internalnorm(z₊ .- z,t)
-    θ = ndz/ndzprev
-    if θ ≥ 1 || ndz * θ^(max_iter - iter) > κtol * (1 - θ)
-      break
+    # compute norm of residuals
+    iter > 1 && (ndzprev = ndz)
+    ndz = integrator.opts.internalnorm(dz, tstep)
+
+    # check divergence (after first iteration)
+    if iter > 1
+      θ = ndz / ndzprev
+      if θ ≥ 1 || ndz * θ^(max_iter - iter) > κtol * (1 - θ)
+        # fixed-point iteration diverges
+        break
+      end
     end
 
-    # update solution
+    # update iterate
     z = z₊
 
     # check stopping criterion
-    η = θ / (1 - θ) # calculated for possible early stopping
-    do_functional = iter < min_iter || η * ndz > κtol
+    iter > 1 && (η = θ / (1 - θ))
+    if η * ndz < κtol && (iter > 1 || !iszero(integrator.success_iter))
+      # fixed-point iteration converges
+      fail_convergence = false
+      break
+    end
+
+    # perform Anderson acceleration
+    if nlcache isa NLAndersonConstantCache && iter < max_iter
+      if iter == aa_start
+        # update cached values for next step of Anderson acceleration
+        dzold = dz
+        z₊old = z₊
+      elseif iter > aa_start
+        # increase size of history
+        history += 1
+
+        # remove oldest history if maximum size is exceeded
+        if history > max_history
+          # circularly shift differences of z₊
+          for i in 1:(max_history-1)
+            Δz₊s[i] = Δz₊s[i + 1]
+          end
+
+          # delete left-most column of QR decomposition
+          qrdelete!(Q, R, max_history)
+
+          # update size of history
+          history = max_history
+        end
+
+        # update history of differences of z₊
+        Δz₊s[history] = @. z₊ - z₊old
+
+        # replace/add difference of residuals as right-most column to QR decomposition
+        qradd!(Q, R, vec(dz .- dzold), history)
+
+        # update cached values
+        dzold = dz
+        z₊old = z₊
+
+        # define current Q and R matrices
+        Qcur, Rcur = view(Q, :, 1:history), UpperTriangular(view(R, 1:history, 1:history))
+
+        # check condition (TODO: incremental estimation)
+        if droptol !== nothing
+          while cond(R) > droptol && history > 1
+            qrdelete!(Q, R, history)
+            history -= 1
+            Qcur, Rcur = view(Q, :, 1:history), UpperTriangular(view(R, 1:history, 1:history))
+          end
+        end
+
+        # solve least squares problem
+        γscur = view(γs, 1:history)
+        ldiv!(Rcur, mul!(γscur, Qcur', vec(dz)))
+        integrator.destats.nsolve += 1
+
+        # update next iterate
+        for i in 1:history
+          z = @. z - γs[i] * Δz₊s[i]
+        end
+
+        # update norm of residuals
+        ndz = integrator.opts.internalnorm(z .- z₊ .+ dz, tstep)
```

Comment:
do we need a temporary?

Reply:
I thought we don't need one in the out-of-place implementation.

Reply:
oh true